gateway/src/apicast/upstream.lua

```diff
@@ -0,0 +1,135 @@
+--- @usage
+--- local upstream = Upstream.new('http://example.com')
+--- upstream:rewrite() -- set Host header to 'example.com'
+--- -- store itself in `context` table for later use in balancer phase and call `ngx.exec`.
+--- upstream:call(context)
+
+local setmetatable = setmetatable
+local tonumber = tonumber
+local str_format = string.format
+
+local resty_resolver = require('resty.resolver')
+local resty_url = require('resty.url')
+local core_base = require('resty.core.base')
+local new_tab = core_base.new_tab
+
+local _M = {
+
+}
+
+local function proxy_pass(upstream)
+    local uri = upstream.uri
+
+    return str_format('%s://%s%s%s%s',
+            uri.scheme,
+            upstream.upstream_name,
+            uri.path or ngx.var.uri or '',
+            ngx.var.is_args or '',
+            ngx.var.query_string or '')
+end
+
+local mt = {
+    __index = _M
+}
+
+
+local function parse_url(url)
+    local parsed, err = resty_url.split(url)
+
+    if err then return nil, err end
+
+    local uri = new_tab(0, 6)
+
+    uri.scheme = parsed[1]
+    uri.user = parsed[2]
+    uri.password = parsed[3]
+    uri.host = parsed[4]
+    uri.port = tonumber(parsed[5])
+    uri.path = parsed[6]
+
+    return uri
+end
+
+
+--- Create new Upstream instance.
+--- @tparam string url
+--- @treturn Upstream|nil upstream instance
+--- @treturn nil|string error when upstream can't be initialized
+--- @static
+function _M.new(url)
+    local uri, err = parse_url(url)
+
+    if err then
+        return nil, 'invalid upstream'
+    end
+
+    return setmetatable({
+        uri = uri,
+        location_name = '@upstream',
+        upstream_name = 'upstream',
+        resolver = resty_resolver,
+    }, mt)
+end
+
+--- Resolve upstream servers.
+--- @treturn {...}|nil resolved servers returned by the resolver
+--- @treturn nil|string error in case resolving fails
+function _M:resolve()
+    local resolver = self.resolver
+    local uri = self.uri
+
+    if self.servers then
+        return self.servers
+    end
+
+    if not resolver or not uri then return nil, 'not initialized' end
+
+    local res, err = resolver:instance():get_servers(uri.host, uri)
+
+    if err then
+        return nil, err
+    end
+
+    self.servers = res
+
+    return res
+end
+
+--- Return port to use when connecting to upstream.
+--- @treturn number port number
+function _M:port()
+    if not self or not self.uri then
+        return nil, 'not initialized'
+    end
+
+    return self.uri.port or resty_url.default_port(self.uri.scheme)
+end
+
+--- Rewrite request Host header to what is provided in the argument or in the URL.
+--- @tparam string host Host header of the request
+function _M:rewrite(host)
+    ngx.req.set_header('Host', host or self.uri.host)
+end
+
+local function exec(self)
+    ngx.var.proxy_pass = proxy_pass(self)
+
+    if self.location_name then
+        ngx.exec(self.location_name)
+    end
+end
+
+--- Execute the upstream.
+function _M:call(context)
```

Comment:
I'm a bit concerned about sharing the policy context with modules that are not policies.
I'll come back to this after I review everything to evaluate some other options.

Reply:
I think other objects can work with context too. In this case it needs to set itself to the context before it calls exec. I think it would make for every weird API and create a lot of duplicity if we would want to control this outside this module.

Reply:
I saw that this is not called only with a policies context, it's also called with `ngx.ctx`.
So context can be any table. Can we document this so we don't forget?